### ISSUE TITLE
Type checking inconsistency with ref and source/active_source

### ISSUE THREAD

**Issue by toots** [savonet/liquidsoap#419]
```
# l = ref [];;
l : ref([?A]) = ref([])
# list.iter(source.shutdown,!l);;
- : unit = ()
# l;;
- : ref([source(?A)]) = ref([])
# l := [output.dummy(blank())];;
At line 1, char 3-4:
  this value has type
    ref([source(_)]) (inferred at line 1, char 5-10)
  but it should be a subtype of (the type of the value at line 1, char 3-28)
    ref([active_source(_)])
```
```
# l = ref [];;
l : ref([?A]) = ref([])
# l' = !l;;
l' : [?A] = []
# l := [output.dummy(blank())];;
- : unit = ()
# l;;
- : ref([active_source(audio=2,video=0,midi=0)]) = ref([<source>])
# l';;
- : [active_source(audio=2,video=0,midi=0)] = []
# list.iter(source.shutdown,l');;
- : unit = ()
```

**Comment by toots** [comment]
Is that really a bug not that I think about it? `active_source` is a sub-type of `source`. Should the final check on the first example restrict the type for the list from `source list` to `active_source list`?

**Comment by dbaelde** [comment]
I don't see any bug here. It's just that type inference isn't always the smartest wrt subtyping, which is to some extend unavoidable.

In the first example, there should be two ways to fix the type inference failure: either a type annotation in one of the first two lines to tell the type system that we want a list of active sources, or a type annotation/coercion in the last line to help the type checker see that the active source list is also a source list and can be stored in the reference. I don't have a working liquidsoap at hand to test that, though.

**Comment by toots** [comment]
I'm taking this one off the bucket list for `1.4.0`, will tackle in during the next major release cycle.

**Comment by smimram** [comment]
I guess this is solved by #1012 .

**Comment by toots** [comment]
Good question. Let me think about it.
